### ISSUE TITLE
adaptive weighting code

### ISSUE THREAD

**Issue by CMakey** [thanif/Camera-Calibration-through-Camera-Projection-Loss#10]
thanks for u impressing work !
But I have some questions about code and wondering for u reply.

1. I'm new to camera calibration, so I'm confused about the directory about there name rules.

2.After reading <Camera calibration through camera projection loss>, i find u said MTL-CPL-A had used adaptive weight method, but I didn't find any code related to it.😵‍💫

Will be appreciated for u reply🌹

**Comment by thanif** [comment]
Hi,

There is a folder named after MTL-CPL-A which contains the code for the same.

**Comment by CMakey** [comment]
thanks for u reply, I'm confused about thise equation ,i can't find any code about it in MTL-CPL-A -> exp-18.ipynb 
![image](https://github.com/thanif/CPL/assets/39455147/4739182d-1a25-461a-b32e-68a4133357a8)

**Comment by thanif** [comment]
Hi,

The equation states that there is a weight assigned to each variable. In exp-18, following code has weights to the variables.

# w_xworld
dense_31 = Dense(120, activation = 'relu')(concat)
dense_32 = Dense(84, activation = 'relu')(dense_31)
w_xworld = Dense(1, name='w_xworld', activation = 'sigmoid')(dense_32)

# w_yworld
dense_33 = Dense(120, activation = 'relu')(concat)
dense_34 = Dense(84, activation = 'relu')(dense_33)
w_yworld = Dense(1, name='w_yworld', activation = 'sigmoid')(dense_34)

# w_zworld
dense_35 = Dense(120, activation = 'relu')(concat)
dense_36 = Dense(84, activation = 'relu')(dense_35)
w_zworld = Dense(1, name='w_zworld', activation = 'sigmoid')(dense_36)

**Comment by CMakey** [comment]
thanks!

**Comment by CMakey** [comment]
A> Hi,
> 
> The equation states that there is a weight assigned to each variable. In exp-18, following code has weights to the variables.
> 
> # w_xworld
> dense_31 = Dense(120, activation = 'relu')(concat) dense_32 = Dense(84, activation = 'relu')(dense_31) w_xworld = Dense(1, name='w_xworld', activation = 'sigmoid')(dense_32)
> 
> # w_yworld
> dense_33 = Dense(120, activation = 'relu')(concat) dense_34 = Dense(84, activation = 'relu')(dense_33) w_yworld = Dense(1, name='w_yworld', activation = 'sigmoid')(dense_34)
> 
> # w_zworld
> dense_35 = Dense(120, activation = 'relu')(concat) dense_36 = Dense(84, activation = 'relu')(dense_35) w_zworld = Dense(1, name='w_zworld', activation = 'sigmoid')(dense_36)

Hello, sorry to bother u again...  I'm still confused about this equation...

In this equation, some adaptive weights are attached to each target loss,  and I can easily find these loss fromcode 
`model = Model(inputs=[left_img, right_img], outputs=[pred_fx, pred_fy, pred_u0, pred_v0, pred_baseline, pred_disparity, pred_x, pred_y, pred_z, pred_pitch, pred_xWorld,pred_yWorld,pred_zWorld])
`

All outputs are predicted for camera's parameters. ( related to w_hat in paper <CAMERA CALIBRATION THROUGH CAMERA PROJECTION LOSS>).
And  `model.compile(loss="mae", target_tensors=[target1, target2, target3, target4, target5, target6, target7, target8, target9, target10, target11, target12, target13],optimizer=optimizers.Adam(lr=learning_rate))
` is related to eq(10).

In order to disentangle the loss, 13 targets are real camera parameters.
Code  `# compile model
model.compile(loss="mae", target_tensors=[target1, target2, target3, target4, target5, target6, target7, target8, target9, target10, target11, target12, target13],optimizer=optimizers.Adam(lr=learning_rate))`is related with eq(11).

Code `    error[i] /= np.shape(output)[1]` is related to normalization in eq(11), 

 But I didn't see the how adaptive weights is used from code...
 
 I'll be appreciated for u reply...

**Comment by thanif** [comment]
Hi,

I don't understand your question.

**Comment by CMakey** [comment]
> Hi,
> 
> I don't understand your question.

Thanks for u reply!!!🌹🌹
sorry, maybe i don't describe my question precisely .
My question is I can't find the code that u used to creative adaptive weights...
the answer u provided before in my comprehension is , u are trying to create three dense layers to compute reprojection loss. 

But I don't understand which code is related to symbol 'alpha'

**Comment by thanif** [comment]
# w_xworld
dense_31 = Dense(120, activation = 'relu')(concat)
dense_32 = Dense(84, activation = 'relu')(dense_31)
w_xworld = Dense(1, name='w_xworld', activation = 'sigmoid')(dense_32)

# w_yworld
dense_33 = Dense(120, activation = 'relu')(concat)
dense_34 = Dense(84, activation = 'relu')(dense_33)
w_yworld = Dense(1, name='w_yworld', activation = 'sigmoid')(dense_34)

# w_zworld
dense_35 = Dense(120, activation = 'relu')(concat)
dense_36 = Dense(84, activation = 'relu')(dense_35)
w_zworld = Dense(1, name='w_zworld', activation = 'sigmoid')(dense_36)

w_xworld is alpha for x_world, w_yworld is alpha for y_world, w_zworld is alpha for z_world

**Comment by CMakey** [comment]
Sorry to bother you again…

Maybe I didn’t explain my question properly…
The question is that in paper eq(12) u calculate the error with 13 alpha weight explicitly , but in code i didn’t see these alpha explicitly. 

I guess maybe in keras ,fit method can automatically adjust these weights ,so u needn't write code for them. 
But after asking GPT,  it showed there should be more code about defining a custom loss function：

> In the context described, the overall loss function \( L^* \) in the paper is a weighted sum of individual task losses \( L_{f_x}, L_{f_y}, \ldots, L_Z \) with weights \( \alpha_{f_x}, \alpha_{f_y}, \ldots \) adjusting the importance of each task loss. In Keras, this kind of weighting adjustment typically isn't automatic. You'd need to define a custom loss function to implement this weighted mechanism and use it during model compilation. While the `fit` method automatically handles loss calculation and backpropagation, the adjustment of weights among different task losses in multi-task learning requires manual setup.


I’m also new to keras , and I wonder if there are any errors of my comprehension...

**Comment by thanif** [comment]
Hi,

As explained earlier, there are weights assigned to xworld, yworld, zworld which act as alphas.

**Comment by CMakey** [comment]
emmm... thanks for u again!
In paper I had seen that u used X, Y, Z to represent camera parameters.

So, do u mean that w_xworld is an proxy weight which represent for several weights? 

After giving all my appreciate, I apology for my ignorance.

**Comment by thanif** [comment]
Hi,

No, w_xworld represents weight only for xworld. Only xworld, yworld, zworld have weights, all other weights are supposed to be 1.
